# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [127]:
import pandas as pd
import numpy as np

# Function to display Dframes side by side
from IPython.display import display_html
def display_side_by_side(*args):
    html_str=''
    for df in args:
        html_str+=df.to_html()
    display_html(html_str.replace('table','table style="display:inline"'),raw=True)


## imports for Python, Pandas

In [128]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [129]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

In [130]:
# use normalization to create tables from nested element
json_normalize(data, 'counties')

name  population
0        Dade       12345
1     Broward       40000
2  Palm Beach       60000
3      Summit        1234
4    Cuyahoga        1337

In [131]:
# further populate tables created from nested element
json_normalize(data, 'counties', ['state', 'shortname', ['info', 'governor']])

name  population    state shortname info.governor
0        Dade       12345  Florida        FL    Rick Scott
1     Broward       40000  Florida        FL    Rick Scott
2  Palm Beach       60000  Florida        FL    Rick Scott
3      Summit        1234     Ohio        OH   John Kasich
4    Cuyahoga        1337     Ohio        OH   John Kasich

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [132]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

In [133]:
# load as Pandas dataframe
sample_json_df = pd.read_json('data/world_bank_projects_less.json')
sample_json_df

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   

  countrycode                              countryname countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia         Ethiopia   
1          TN                      Republic of Tunisia          Tunisia   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   

                                      theme_namecode themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]        65  130000000   
1  [{'name': 'Other economic management', 'code':...     54,24          0   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  

[2 rows x 50 columns]

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

## 1. Find the 10 countries with most projects

In [134]:
# import the json file assign it to a variable and determine the column names.

json_df = pd.read_json('data/world_bank_projects.json')
json_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

In [135]:
# Extract subset of the data with by country name and project name.

df1 = json_df[['countryname','countrycode','project_name' ]]
df1.head()

countryname countrycode  \
0  Federal Democratic Republic of Ethiopia          ET   
1                      Republic of Tunisia          TN   
2                                   Tuvalu          TV   
3                        Republic of Yemen          RY   
4                       Kingdom of Lesotho          LS   

                                        project_name  
0  Ethiopia General Education Quality Improvement...  
1          TN: DTF Social Protection Reforms Support  
2  Tuvalu Aviation Investment Project - Additiona...  
3   Gov't and Civil Society Organization Partnership  
4  Second Private Sector Competitiveness and Econ...

In [136]:
# inspect information of df1 to determine if there are any null values.
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 3 columns):
countryname     500 non-null object
countrycode     500 non-null object
project_name    500 non-null object
dtypes: object(3)
memory usage: 15.6+ KB


In [137]:
#Determine if there are any fields with just a blank space since pandas recognizes blank space as an object and not as nan
nan_count ={}
clist = ['countryname','countrycode','project_name' ]

for x in clist:
    nan_count[x] = json_df[json_df[x] == '']

for k, v in nan_count.items():
    print(k,' : ', len(v))

countryname  :  0
countrycode  :  0
project_name  :  0


In [138]:
# Count the number of projects per country by grouping by country, 
# sorting the values by project in descending order and considering the first 10 with head(10)
# Could equally use pivo_tables as shown

# proj = df1.pivot_table(index=['countryname','countrycode'], values = 'project_name', aggfunc='count').sort_values('project_name', ascending=False).head(10)
# proj

proj_count= df1.groupby(['countryname', 'countrycode']).count().sort_values('project_name', ascending=False).head(10)
proj_count



project_name
countryname                     countrycode              
People's Republic of China      CN                     19
Republic of Indonesia           ID                     19
Socialist Republic of Vietnam   VN                     17
Republic of India               IN                     16
Republic of Yemen               RY                     13
People's Republic of Bangladesh BD                     12
Nepal                           NP                     12
Kingdom of Morocco              MA                     12
Republic of Mozambique          MZ                     11
Africa                          3A                     11

* The top 10 countries are as listed above.
* Africa is not a country but from the data, some projects were carried out in Africa by multiple countries.

## 2. Find the top 10 major project themes

In [139]:
#visualize the dataframe

json_df[['countryname','countrycode','mjtheme_namecode']].head(20)


countryname countrycode  \
0   Federal Democratic Republic of Ethiopia          ET   
1                       Republic of Tunisia          TN   
2                                    Tuvalu          TV   
3                         Republic of Yemen          RY   
4                        Kingdom of Lesotho          LS   
5                         Republic of Kenya          KE   
6                         Republic of India          IN   
7                People's Republic of China          CN   
8                         Republic of India          IN   
9                        Kingdom of Morocco          MA   
10                  Republic of South Sudan          SS   
11                        Republic of India          IN   
12                        Republic of Ghana          GH   
13       Democratic Republic of Timor-Leste          TP   
14              Hashemite Kingdom of Jordan          JO   
15                                    Samoa          WS   
16                                    Samoa          WS   
17                   Republic of Madagascar          MG   
18                      Kingdom of Cambodia          KH   
19                       Kingdom of Morocco          MA   

                                     mjtheme_namecode  
0   [{'code': '8', 'name': 'Human development'}, {...  
1   [{'code': '1', 'name': 'Economic management'},...  
2   [{'code': '5', 'name': 'Trade and integration'...  
3   [{'code': '7', 'name': 'Social dev/gender/incl...  
4   [{'code': '5', 'name': 'Trade and integration'...  
5   [{'code': '6', 'name': 'Social protection and ...  
6   [{'code': '2', 'name': 'Public sector governan...  
7   [{'code': '11', 'name': 'Environment and natur...  
8   [{'code': '10', 'name': 'Rural development'}, ...  
9   [{'code': '2', 'name': 'Public sector governan...  
10  [{'code': '10', 'name': 'Rural development'}, ...  
11  [{'code': '10', 'name': 'Rural development'}, ...  
12                        [{'code': '4', 'name': ''}]  
13  [{'code': '5', 'name': 'Trade and integration'...  
14  [{'code': '6', 'name': 'Social protection and ...  
15  [{'code': '10', 'name': 'Rural development'}, ...  
16  [{'code': '10', 'name': 'Rural development'}, ...  
17  [{'code': '8', 'name': 'Human development'}, {...  
18  [{'code': '8', 'name': 'Human development'}, {...  
19  [{'code': '2', 'name': 'Public sector governan...

In [140]:
# Load json file and use json_normalize on just the project theme column which is in itself a dictionary and 
# look at the first 10 rows

json_data = json.load(open('data/world_bank_projects.json'))
json_strdf = json_normalize(json_data, 'mjtheme_namecode')
json_strdf.head(20)

code                                          name
0     8                             Human development
1    11                                              
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration
5     2                      Public sector governance
6    11  Environment and natural resources management
7     6         Social protection and risk management
8     7                   Social dev/gender/inclusion
9     7                   Social dev/gender/inclusion
10    5                         Trade and integration
11    4      Financial and private sector development
12    6         Social protection and risk management
13    6                                              
14    2                      Public sector governance
15    4      Financial and private sector development
16   11  Environment and natural resources management
17    8                                              
18   10                             Rural development
19    7

In [141]:
#Determine if there are any fields with just the blank space.

space_count ={}
clist = ['code','name' ]

for x in clist:
    space_count[x] = json_strdf[json_strdf[x] == '']

for k, v in space_count.items():
    print(k,' : ', len(v))



code  :  0
name  :  122


Clearly there are no blank spaces on the code column but there are 122 on the name column. 
so if we group by code we should still have an accurate count of the project theme names.

In [142]:
# Groupby code, count and sort the number of occurences.

code_count= json_strdf.groupby('code').count().sort_values('name', ascending=False)
code_count

name
code      
11     250
10     216
8      210
2      199
6      168
4      146
7      130
5       77
9       50
1       38
3       15

In [143]:
#Create Unique dataframe of code and code names

proj_lst = json_strdf.replace('', np.nan).drop_duplicates().dropna().reset_index(drop =True)
proj_lst

code                                          name
0     8                             Human development
1     1                           Economic management
2     6         Social protection and risk management
3     5                         Trade and integration
4     2                      Public sector governance
5    11  Environment and natural resources management
6     7                   Social dev/gender/inclusion
7     4      Financial and private sector development
8    10                             Rural development
9     9                             Urban development
10    3                                   Rule of law

In [144]:
# set code as new index
proj_theme = proj_lst.set_index('code')
proj_theme

name
code                                              
8                                Human development
1                              Economic management
6            Social protection and risk management
5                            Trade and integration
2                         Public sector governance
11    Environment and natural resources management
7                      Social dev/gender/inclusion
4         Financial and private sector development
10                               Rural development
9                                Urban development
3                                      Rule of law

In [160]:
# Merge two dataframes to show theme names
proj_theme_most = pd.merge(code_count,proj_theme, left_index=True, right_index = True, suffixes=('-count','-theme'))
proj_theme_most

name-count                                    name-theme
code                                                          
11           250  Environment and natural resources management
10           216                             Rural development
8            210                             Human development
2            199                      Public sector governance
6            168         Social protection and risk management
4            146      Financial and private sector development
7            130                   Social dev/gender/inclusion
5             77                         Trade and integration
9             50                             Urban development
1             38                           Economic management
3             15                                   Rule of law

**The project theme with the highest count is "Environmental and natural resources management"**

## Alternative Method

In [146]:
json_df.mjtheme_namecode.head(10)

0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
2    [{'code': '5', 'name': 'Trade and integration'...
3    [{'code': '7', 'name': 'Social dev/gender/incl...
4    [{'code': '5', 'name': 'Trade and integration'...
5    [{'code': '6', 'name': 'Social protection and ...
6    [{'code': '2', 'name': 'Public sector governan...
7    [{'code': '11', 'name': 'Environment and natur...
8    [{'code': '10', 'name': 'Rural development'}, ...
9    [{'code': '2', 'name': 'Public sector governan...
Name: mjtheme_namecode, dtype: object

In [161]:
# count the number of occurences of each code value.

d_count={}
for i in json_df.mjtheme_namecode:
    for j in i:
        code=j['code']
        name=j['name']
        d_count[code]= d_count.get(code,0)+1       
print(d_count)



{'8': 210, '11': 250, '1': 38, '6': 168, '5': 77, '2': 199, '7': 130, '4': 146, '10': 216, '9': 50, '3': 15}


In [148]:
# Create a unique list of codes and theme names
uniq_list={}
for i in json_df.mjtheme_namecode:
    for j in i:
        code=j['code']
        name=j['name']
        if (code not in uniq_list) and (name!=''):
            uniq_list[code] = name
print(uniq_list)

{'8': 'Human development', '1': 'Economic management', '6': 'Social protection and risk management', '5': 'Trade and integration', '2': 'Public sector governance', '11': 'Environment and natural resources management', '7': 'Social dev/gender/inclusion', '4': 'Financial and private sector development', '10': 'Rural development', '9': 'Urban development', '3': 'Rule of law'}


In [162]:
# Create a dictionary of codes, theme names and number of code occurencies

new_list = []

for k,v in d_count.items():
    new_dict = {}
    new_dict['code'] = k
    new_dict['name'] = uniq_list[k]
    new_dict['theme_count'] = v
    new_list.append(new_dict)

df_theme_count = pd.DataFrame.from_dict(new_list).sort_values('theme_count', ascending=False)

df_theme_count.set_index('code')



name  theme_count
code                                                           
11    Environment and natural resources management          250
10                               Rural development          216
8                                Human development          210
2                         Public sector governance          199
6            Social protection and risk management          168
4         Financial and private sector development          146
7                      Social dev/gender/inclusion          130
5                            Trade and integration           77
9                                Urban development           50
1                              Economic management           38
3                                      Rule of law           15

## 3. Some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [150]:
# Replace missing names with NaN.
js_df_na=json_strdf.replace('', np.nan)
js_df_na.head(20)

code                                          name
0     8                             Human development
1    11                                           NaN
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration
5     2                      Public sector governance
6    11  Environment and natural resources management
7     6         Social protection and risk management
8     7                   Social dev/gender/inclusion
9     7                   Social dev/gender/inclusion
10    5                         Trade and integration
11    4      Financial and private sector development
12    6         Social protection and risk management
13    6                                           NaN
14    2                      Public sector governance
15    4      Financial and private sector development
16   11  Environment and natural resources management
17    8                                           NaN
18   10                             Rural development
19    7                                           NaN

In [151]:
# Function to display Dframes side by side
from IPython.display import display_html
def display_side_by_side(*args):
    html_str=''
    for df in args:
        html_str+=df.to_html()
    display_html(html_str.replace('table','table style="display:inline"'),raw=True)


# METHOD 1: CHANGING CODE COLUMN TO INT, SORTING AND THEN FFILL WITH A FOR LOOP AND APPENDING 

js_df_na['code'] = js_df_na['code'].astype('int64')

js_df_sort = js_df_na.sort_values('code', ascending=False) # Sort by the code which are now int

js_nonans1 = pd.DataFrame() # create and empty dataframe

for k in range(11): # loop over each code item, fill forward and append to empty dataframe created
    js_nonans1 = js_nonans1.append(js_df_sort.loc[js_df_sort['code']== (k+1)].fillna(method='ffill'))
    
display_side_by_side(js_nonans1.sort_index().head(20),js_nonans1.sort_index().tail(20))

code 
 name 
 
 
 
 
 0 
 8 
 Human development 
 
 
 1 
 11 
 Environment and natural resources management 
 
 
 2 
 1 
 Economic management 
 
 
 3 
 6 
 Social protection and risk management 
 
 
 4 
 5 
 Trade and integration 
 
 
 5 
 2 
 Public sector governance 
 
 
 6 
 11 
 Environment and natural resources management 
 
 
 7 
 6 
 Social protection and risk management 
 
 
 8 
 7 
 Social dev/gender/inclusion 
 
 
 9 
 7 
 Social dev/gender/inclusion 
 
 
 10 
 5 
 Trade and integration 
 
 
 11 
 4 
 Financial and private sector development 
 
 
 12 
 6 
 Social protection and risk management 
 
 
 13 
 6 
 Social protection and risk management 
 
 
 14 
 2 
 Public sector governance 
 
 
 15 
 4 
 Financial and private sector development 
 
 
 16 
 11 
 Environment and natural resources management 
 
 
 17 
 8 
 Human development 
 
 
 18 
 10 
 Rural development 
 
 
 19 
 7 
 Social dev/gender/inclusion 
 
 
 
 
 
 
 code 
 name 
 
 
 
 
 1479 
 7 
 Social dev/gender/inclusion 
 
 
 1480 
 11 
 Environment and natural resources management 
 
 
 1481 
 5 
 Trade and integration 
 
 
 1482 
 6 
 Social protection and risk management 
 
 
 1483 
 8 
 Human development 
 
 
 1484 
 4 
 Financial and private sector development 
 
 
 1485 
 7 
 Social dev/gender/inclusion 
 
 
 1486 
 8 
 Human development 
 
 
 1487 
 5 
 Trade and integration 
 
 
 1488 
 2 
 Public sector governance 
 
 
 1489 
 8 
 Human development 
 
 
 1490 
 10 
 Rural development 
 
 
 1491 
 6 
 Social protection and risk management 
 
 
 1492 
 10 
 Rural development 
 
 
 1493 
 10 
 Rural development 
 
 
 1494 
 10 
 Rural development 
 
 
 1495 
 9 
 Urban development 
 
 
 1496 
 8 
 Human development 
 
 
 1497 
 5 
 Trade and integration 
 
 
 1498 
 4 
 Financial and private sector development

In [152]:
# METHOD 2: USING MERGE OF UNIQUE DFRAME OF THEMES AND CODES (proj_lst) AND LIST OF ALL THEMES AND CODES INCLUDING NaNs
# (js_df_na) AND THEN DROPPING THE COLUMN THAT HAS NaNs.

js_df_na2=json_strdf.replace('', np.nan)
js_nonans2 = proj_lst.merge(js_df_na2,on='code', how='left', suffixes=('', '_with Nans'))

display_side_by_side(js_nonans2.head(20),js_nonans2.tail(20))

code 
 name 
 name_with Nans 
 
 
 
 
 0 
 8 
 Human development 
 Human development 
 
 
 1 
 8 
 Human development 
 NaN 
 
 
 2 
 8 
 Human development 
 Human development 
 
 
 3 
 8 
 Human development 
 Human development 
 
 
 4 
 8 
 Human development 
 Human development 
 
 
 5 
 8 
 Human development 
 Human development 
 
 
 6 
 8 
 Human development 
 Human development 
 
 
 7 
 8 
 Human development 
 Human development 
 
 
 8 
 8 
 Human development 
 Human development 
 
 
 9 
 8 
 Human development 
 Human development 
 
 
 10 
 8 
 Human development 
 Human development 
 
 
 11 
 8 
 Human development 
 Human development 
 
 
 12 
 8 
 Human development 
 NaN 
 
 
 13 
 8 
 Human development 
 Human development 
 
 
 14 
 8 
 Human development 
 Human development 
 
 
 15 
 8 
 Human development 
 Human development 
 
 
 16 
 8 
 Human development 
 Human development 
 
 
 17 
 8 
 Human development 
 Human development 
 
 
 18 
 8 
 Human development 
 Human development 
 
 
 19 
 8 
 Human development 
 Human development 
 
 
 
 
 
 
 code 
 name 
 name_with Nans 
 
 
 
 
 1479 
 9 
 Urban development 
 Urban development 
 
 
 1480 
 9 
 Urban development 
 Urban development 
 
 
 1481 
 9 
 Urban development 
 Urban development 
 
 
 1482 
 9 
 Urban development 
 Urban development 
 
 
 1483 
 9 
 Urban development 
 Urban development 
 
 
 1484 
 3 
 Rule of law 
 Rule of law 
 
 
 1485 
 3 
 Rule of law 
 Rule of law 
 
 
 1486 
 3 
 Rule of law 
 Rule of law 
 
 
 1487 
 3 
 Rule of law 
 NaN 
 
 
 1488 
 3 
 Rule of law 
 Rule of law 
 
 
 1489 
 3 
 Rule of law 
 Rule of law 
 
 
 1490 
 3 
 Rule of law 
 NaN 
 
 
 1491 
 3 
 Rule of law 
 Rule of law 
 
 
 1492 
 3 
 Rule of law 
 NaN 
 
 
 1493 
 3 
 Rule of law 
 Rule of law 
 
 
 1494 
 3 
 Rule of law 
 Rule of law 
 
 
 1495 
 3 
 Rule of law 
 Rule of law 
 
 
 1496 
 3 
 Rule of law 
 Rule of law 
 
 
 1497 
 3 
 Rule of law 
 Rule of law 
 
 
 1498 
 3 
 Rule of law 
 Rule of law

In [153]:
js_nonans2=js_nonans2.drop(columns ='name_with Nans')
display_side_by_side(js_nonans2.head(20),js_nonans2.tail(20))

code 
 name 
 
 
 
 
 0 
 8 
 Human development 
 
 
 1 
 8 
 Human development 
 
 
 2 
 8 
 Human development 
 
 
 3 
 8 
 Human development 
 
 
 4 
 8 
 Human development 
 
 
 5 
 8 
 Human development 
 
 
 6 
 8 
 Human development 
 
 
 7 
 8 
 Human development 
 
 
 8 
 8 
 Human development 
 
 
 9 
 8 
 Human development 
 
 
 10 
 8 
 Human development 
 
 
 11 
 8 
 Human development 
 
 
 12 
 8 
 Human development 
 
 
 13 
 8 
 Human development 
 
 
 14 
 8 
 Human development 
 
 
 15 
 8 
 Human development 
 
 
 16 
 8 
 Human development 
 
 
 17 
 8 
 Human development 
 
 
 18 
 8 
 Human development 
 
 
 19 
 8 
 Human development 
 
 
 
 
 
 
 code 
 name 
 
 
 
 
 1479 
 9 
 Urban development 
 
 
 1480 
 9 
 Urban development 
 
 
 1481 
 9 
 Urban development 
 
 
 1482 
 9 
 Urban development 
 
 
 1483 
 9 
 Urban development 
 
 
 1484 
 3 
 Rule of law 
 
 
 1485 
 3 
 Rule of law 
 
 
 1486 
 3 
 Rule of law 
 
 
 1487 
 3 
 Rule of law 
 
 
 1488 
 3 
 Rule of law 
 
 
 1489 
 3 
 Rule of law 
 
 
 1490 
 3 
 Rule of law 
 
 
 1491 
 3 
 Rule of law 
 
 
 1492 
 3 
 Rule of law 
 
 
 1493 
 3 
 Rule of law 
 
 
 1494 
 3 
 Rule of law 
 
 
 1495 
 3 
 Rule of law 
 
 
 1496 
 3 
 Rule of law 
 
 
 1497 
 3 
 Rule of law 
 
 
 1498 
 3 
 Rule of law

**js_nonans is a dataframe with all the "NaN" values  filled.**

In [154]:
# Determine that there were no overlaps during the forward fill process of code names 
display_side_by_side(js_nonans1.groupby('code').count().sort_values('name', ascending=False),js_nonans2.groupby('code').count().sort_values('name', ascending=False))

name 
 
 
 code 
 
 
 
 
 
 11 
 250 
 
 
 10 
 216 
 
 
 8 
 210 
 
 
 2 
 199 
 
 
 6 
 168 
 
 
 4 
 146 
 
 
 7 
 130 
 
 
 5 
 77 
 
 
 9 
 50 
 
 
 1 
 38 
 
 
 3 
 14 
 
 
 
 
 
 
 name 
 
 
 code 
 
 
 
 
 
 11 
 250 
 
 
 10 
 216 
 
 
 8 
 210 
 
 
 2 
 199 
 
 
 6 
 168 
 
 
 4 
 146 
 
 
 7 
 130 
 
 
 5 
 77 
 
 
 9 
 50 
 
 
 1 
 38 
 
 
 3 
 15

**We get the same values as initially determined before the forward fill hence no overlaps.**

In [155]:
# METHOD 3: USING A FUNCTION TO FILL IN THE MISSING VALUES WITHIN THE DATAFRAME

json_df.mjtheme_namecode.head(20)

0     [{'code': '8', 'name': 'Human development'}, {...
1     [{'code': '1', 'name': 'Economic management'},...
2     [{'code': '5', 'name': 'Trade and integration'...
3     [{'code': '7', 'name': 'Social dev/gender/incl...
4     [{'code': '5', 'name': 'Trade and integration'...
5     [{'code': '6', 'name': 'Social protection and ...
6     [{'code': '2', 'name': 'Public sector governan...
7     [{'code': '11', 'name': 'Environment and natur...
8     [{'code': '10', 'name': 'Rural development'}, ...
9     [{'code': '2', 'name': 'Public sector governan...
10    [{'code': '10', 'name': 'Rural development'}, ...
11    [{'code': '10', 'name': 'Rural development'}, ...
12                          [{'code': '4', 'name': ''}]
13    [{'code': '5', 'name': 'Trade and integration'...
14    [{'code': '6', 'name': 'Social protection and ...
15    [{'code': '10', 'name': 'Rural development'}, ...
16    [{'code': '10', 'name': 'Rural development'}, ...
17    [{'code': '8', 'name': 'Human development'

 The code 4 on line 12 above is missing a name.

In [164]:
# Count number of blank spaces

pst = {}
nst=0
for i in json_df.mjtheme_namecode:
    for j in i:
        code=j['code']
        name=j['name']
        if (code !='') and (name ==''):
             pst[code]= pst.get(code,0)+1 
                
for k, v in pst.items():
    
    nst= nst + v
    
print(pst)    
print('There are ', nst, ' spaces')

{'11': 27, '6': 10, '8': 13, '7': 11, '2': 15, '4': 16, '10': 14, '5': 5, '1': 5, '9': 3, '3': 3}
There are  122  spaces


In [157]:
# Create a function to fill missing values.

test_list=[{'code': '1', 'name': ''},{'code': '6', 'name': 'Social protection and risk management'}]

def impute_missing(list):
    output=[]
    for i in list:
        if i['name'] =='':
            temp_dict={}
            name = uniq_list[i['code']]
            temp_dict['code']=i['code']
            temp_dict['name']=name
            output.append(temp_dict)
        else:
            output.append(i)
    return output

impute_missing(test_list)

[{'code': '1', 'name': 'Economic management'},
 {'code': '6', 'name': 'Social protection and risk management'}]

In [158]:
# apply impute_missing function we see that

json_ffill = json_df.mjtheme_namecode.apply(impute_missing)
json_ffill

0      [{'code': '8', 'name': 'Human development'}, {...
1      [{'code': '1', 'name': 'Economic management'},...
2      [{'code': '5', 'name': 'Trade and integration'...
3      [{'code': '7', 'name': 'Social dev/gender/incl...
4      [{'code': '5', 'name': 'Trade and integration'...
5      [{'code': '6', 'name': 'Social protection and ...
6      [{'code': '2', 'name': 'Public sector governan...
7      [{'code': '11', 'name': 'Environment and natur...
8      [{'code': '10', 'name': 'Rural development'}, ...
9      [{'code': '2', 'name': 'Public sector governan...
10     [{'code': '10', 'name': 'Rural development'}, ...
11     [{'code': '10', 'name': 'Rural development'}, ...
12     [{'code': '4', 'name': 'Financial and private ...
13     [{'code': '5', 'name': 'Trade and integration'...
14     [{'code': '6', 'name': 'Social protection and ...
15     [{'code': '10', 'name': 'Rural development'}, ...
16     [{'code': '10', 'name': 'Rural development'}, ...
17     [{'code': '8', 'name': '

The code 4 on line 12 has been filled.

In [159]:
# Testing for empty spaces
pst = {}
for i in json_ffill:
    for j in i:
        name=j['name']
        if name =='':
             pst[name]= pst.get(name,0)+1 

print(pst)


{}


**All the spaces have been filled with names**